sycl/include/sycl/ext/oneapi/experimental/builtins.hpp

```diff
@@ -117,6 +118,152 @@ inline __SYCL_ALWAYS_INLINE
 
 } // namespace native
 
+namespace detail {
+
+template <typename T> struct is_bf16_storage_type {
+  static constexpr int value = false;
+};
+
+template <> struct is_bf16_storage_type<uint16_t> {
+  static constexpr int value = true;
+};
+
+template <> struct is_bf16_storage_type<uint32_t> {
+  static constexpr int value = true;
+};
+
+template <int N> struct is_bf16_storage_type<vec<uint16_t, N>> {
+  static constexpr int value = true;
+};
+
+template <int N> struct is_bf16_storage_type<vec<uint32_t, N>> {
+  static constexpr int value = true;
+};
+
+} // namespace detail
+
+template <typename T>
+std::enable_if_t<experimental::detail::is_bf16_storage_type<T>::value, T>
+fabs(T x) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return __clc_fabs(x);
+#else
+  (void)x;
+  throw runtime_error("bfloat16 is not currently supported on the host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+template <typename T>
+std::enable_if_t<std::is_same<T, bfloat16>::value, T> fabs(T x) {
```

Comment:
Modern C++ everywhere
```suggestion
std::enable_if_t<std::is_same_t<T, bfloat16>, T> fabs(T x) {
```

Reply:
Thanks, done.

Reply:
It fails a test due to this: I think you have to add a flag for c++17, I think this is why I avoided it previously.  I'll investigate further.

Reply:
SYCL requires at least C++17.

Reply:
> SYCL requires at least C++17.

Though that is true for SYCL 2020, DPC++ currently attempts to expose as much of the API as possible while only causing warnings when C++14 is used. See https://github.com/intel/llvm/blob/sycl/sycl/test/basic_tests/stdcpp_compat.cpp.

It is allowed to have these only defined for C++17, in which case it can use C++17 features, but if it is simply for this purpose I believe it is preferred that the [`detail` variants](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/stl_type_traits.hpp) of C++17 features are used for now.

Reply:
> > SYCL requires at least C++17.
> 
> Though that is true for SYCL 2020, DPC++ currently attempts to expose as much of the API as possible while only causing warnings when C++14 is used. See https://github.com/intel/llvm/blob/sycl/sycl/test/basic_tests/stdcpp_compat.cpp.
> 
> It is allowed to have these only defined for C++17, in which case it can use C++17 features, but if it is simply for this purpose I believe it is preferred that the [`detail` variants](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/stl_type_traits.hpp) of C++17 features are used for now.

Is using `sycl::detail` variants of C++17 syntax better than using the C++14 syntax for features in which the C++14 and C++17 syntax is functionally identical and can be switched as easily in the future as today?
Presumably at some point we will want to switch to `std::` C++17. 
Is it not potentially confusing to the uninitiated that there are `sycl::detail` variants of C++17? If the C++17 feature is useful, and is not available in C++14 or is implemented in a functionally different way to in C++14, I see that the usage of `sycl::detail` C++17 now is a good idea. 

For `std::is_same<T, bfloat16>::value` vs `sycl::detail::is_same_t<T, bfloat16>` I don't see the advantage from what I know.

Reply:
In general I agree, for shorthands the advantage is negligible, but I think there are the following benefits:

1. Brevity, i.e. `detail::is_same_v<T, bfloat16>` is slightly shorter than `std::is_same<T, bfloat16>::value` (though not by much.) However, for this you are right that since you're in an extension and there may be some ambiguity about the `detail` namespace, you likely need `sycl::` too so it becomes a longer line.
2. Switching to C++17, we can simply remove the `detail` definitions and that will give us errors showing all the places using these, so we can easily go replace `detail` with `std` and everything should be good and ready. But again, there's nothing wrong with using things like `std::is_same<...>::value` even with the shorthands available, so it's not a make-or-break situation.

Whether or not you want to use the `detail` definitions I'll leave up to you, but the important note here is that we should not be using C++17 features in the headers currently.

Reply:
Thanks, I used `sycl::detail::is_same_v` initially but it led to test failures so I'm back to C++14 now!